### PR TITLE
feat(pkg/mmr): Add basic MMR support

### LINKED ISSUE #4152: (BEEFY): MMR MVP
Create an small MMR mvp to validate the https://hackmd.io/@dimartiro/H1L_R-8iR

### PR DESCRIPTION
**Pull request by dimartiro**
## Changes

Add basic MMR support following the 1st iteration implementation proposed [here](https://hackmd.io/@dimartiro/H1L_R-8iR)

### Features
- [x] Leaves accumulation
- [x] Root calculation
- [x] Commit changes to underlying storage 
- [x] Basic testing  

## Tests

```sh
go test ./pkg/mmr
```

## Issues

Closes: #4152

### CODE REVIEW

File: pkg/mmr/mmr_batch.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+			continue
+		} else if pos < node.pos-uint64(len(node.elements)) {
+			elementPosition := pos - node.pos
+			if elementPosition < uint64(len(node.elements)) {
+				return &node.elements[int(pos-node.pos)], nil
+			}
+			return nil, nil
+		} else {
+			break
+		}
```
Comment (EclesioMeloJunior) [inline pkg/mmr/mmr_batch.go]:
```suggestion
    break
```

**Review by EclesioMeloJunior — APPROVED** [review]
LGTM! Would like to see more tests over getElement, also can the storage be generic over `MMRElement`?

**Comment by dimartiro** [comment]
> Would like to see more tests over getElement

I'm gonna add more, thanks for the review 👍 

> also can the storage be generic over `MMRElement`?

You mean to use a generic instead of using `MMRElement`? basically `MMRElement` is an alias for `[]byte` I wanted to keep it simple since it is an MVP and using `[]byte` is easier for us since we don't need any special serializer or deserializer.
I think we can improve it later if we have any future special requirement to use a generic

File: pkg/mmr/memstorage.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+func (s *MemStorage) append(pos uint64, elements []MMRElement) error {
+	for i, element := range elements {
+		s.storage.Set(pos+uint64(i), element)
+	}
+	return nil
+}
+
+func (s *MemStorage) commit() error {
```
Comment (jimjbrettj) [inline pkg/mmr/memstorage.go]: This is for future functionality yes?

File: pkg/mmr/memstorage.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package mmr
+
+import (
+	"github.com/tidwall/btree"
+)
+
+type MemStorage struct {
```
Comment (jimjbrettj) [inline pkg/mmr/memstorage.go]: I think a little description here would be useful. At least describing the contents of the map (I had to double-check what the key was)

File: pkg/mmr/mmr.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+		peak <<= 1
+		position += 1
+		leftPosition := position - peak
+		leftElement, err := mmr.findElement(leftPosition, elements)
+
+		if err != nil {
+			return 0, err
+		}
+
+		rightElement := elements[len(elements)-1] // TODO: check this wont fail
```
Comment (jimjbrettj) [inline pkg/mmr/mmr.go]: Is this an active todo?

**Review by jimjbrettj — APPROVED** [review]
Really nice work! I already looked over it, but if you could do a once over of the GP appendix E to make sure it seems compatible before merging that would be great.

**Comment by dimartiro** [comment]
> Really nice work! I already looked over it, but if you could do a once over of the GP appendix E to make sure it seems compatible before merging that would be great.

I've been checking, and if I'm not mistaken, it aligns well with the JAM GP.

image

The `Append` function describes how to add a new element in a generic way and how to merge it with the current peaks. If I understand correctly, it considers an MMR, which isn't necessarily limited to binary tries, which is why the function `P` is described based on `n`.

Regarding the encoding function, it essentially describes the process of bagging the peaks and calculating the final MMR root so we are ok with that too.

This is all based on my interpretation of the functions described in the GP, but in the future, we can compare our implementation with Parity’s to ensure we have the same behavior 😄

File: pkg/mmr/mmr.go [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+// retrieving, appending, and committing elements to the MMR structure.
+type MMRStorage[T any] interface {
+	getElement(pos uint64) (*T, error)
+	append(pos uint64, elements []T) error
+	commit() error
+}
+
+// MergeFunc defines a function type used to merge two elements in the MMR.
+// It takes two elements, `left` and `right`, and returns a pointer to the
+// merged result or an error if the merge operation fails.
```
Comment (EclesioMeloJunior) [inline pkg/mmr/mmr.go:27]: I think this comment is not needed, the func signature is self-explanatory imo, maybe you can use it to explain how the merge is implemented.

File: pkg/mmr/mmr.go [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+		return nil, errorInconsistentStore
+	}
+
+	return value, nil
+}
+
+/*
+Returns a bitmap of the peaks in the MMR.
+Eg: 0b11 means that the MMR has 2 peaks at position 0 and at position 1
+*/
```
Comment (EclesioMeloJunior) [inline pkg/mmr/mmr.go]: nit, just to normalize the comments
```suggestion
// Returns a bitmap of the peaks in the MMR.
// Eg: 0b11 means that the MMR has 2 peaks at position 0 and at position 1
```

**Review by EclesioMeloJunior — APPROVED** [review]

File: pkg/mmr/mmr.go (hunk 5 above)
Comment (dimartiro) [reply pkg/mmr/mmr.go:27]: Agree, but I wanted to add a comment since it is an exported type

**Comment by dimartiro** [comment]
Closed in favor of https://github.com/ChainSafe/go-jam/pull/38

**Pull request closed without merging**